File: hash/PG_베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

bool cmp(pair<int, int> p1, pair<int, int> p2) {
    if(p1.first == p2.first) return p1.second < p2.second;
    return p1.first > p2.first;
}

bool great(pair<string, int> p1, pair<string, int> p2) {
    return p1.second > p2.second;
}

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    unordered_map<string, int> gp;
    unordered_map<string, vector<pair<int, int>>> um;
    for(int i = 0; i < genres.size(); i++) {
        um[genres[i]].push_back({plays[i], i});
        gp[genres[i]] += plays[i];
    }
    // map은 pair<key, value> 형태로 저장되어 있어, vector를 pair형으로 선언하고 옮겨담을 수 있다
    vector<pair<string, int>> genre_plays(gp.begin(), gp.end());
    sort(genre_plays.begin(), genre_plays.end(), great);
    for(int i = 0; i < genre_plays.size(); i++) {
        sort(um[genre_plays[i].first].begin(), um[genre_plays[i].first].end(), cmp);
        answer.push_back(um[genre_plays[i].first][0].second);
        if(um[genre_plays[i].first].size() > 1) answer.push_back(um[genre_plays[i].first][1].second);
    }
    return answer;
}

File: hash/PG_베스트앨범.java
import java.util.*;
import java.util.Map.Entry;
import java.io.*;

class Genre implements Comparable<Genre> {
    int music;
    int played;

    Genre() {

    }

    Genre(int music, int played) {
        this.music = music;
        this.played = played;
    }

    public int getMusic() {
        return music;
    }

    public int getPlayed() {
        return played;
    }

    @Override
    public int compareTo(Genre g1) {
        if(g1.getPlayed() == getPlayed()) {
            // 고유번호 -- 오름차순
            return getMusic() - g1.getMusic();
        } else {
            // 노래 재생 횟수 -- 내림차순
            return g1.getPlayed() - getPlayed();
        }
    }
}

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        List<Integer> anslist = new ArrayList<>();
        Map<String, Integer> genreTotal = new HashMap<>();
        Map<String, List<Genre>> genreMusic = new HashMap<>();
        for(int i = 0; i < genres.length; i++) {
            if(genreTotal.containsKey(genres[i])) {
                // 각 장르에 재생횟수 더한 값을 value로
                genreTotal.put(genres[i], genreTotal.get(genres[i])+plays[i]);
            } else {
                genreTotal.put(genres[i], plays[i]);
            }
        }
        for (int i = 0; i < plays.length; i++) {
            if(genreMusic.containsKey(genres[i])) {
                genreMusic.get(genres[i]).add(new Genre(i, plays[i]));
            } else {
                genreMusic.put(genres[i], new ArrayList<>());
                genreMusic.get(genres[i]).add(new Genre(i, plays[i]));
            }
        }
        // Map to List (map의 value값으로 정렬하기 위해)
        List<Entry<String, Integer>> entries = new ArrayList<>(genreTotal.entrySet());
        // value로 내림차순 정렬
        Collections.sort(entries, new Comparator<Entry<String, Integer>>() {
            @Override
            public int compare(Entry<String, Integer> o1, Entry<String, Integer> o2) {
                return o2.getValue().compareTo(o1.getValue());
            }
        });
        for(String key : genreMusic.keySet()) {
            // 객체 리스트 정렬
            Collections.sort(genreMusic.get(key));
        }
        int i = 0;
        for(Entry<String, Integer> entry : entries) {
            anslist.add(genreMusic.get(entry.getKey()).get(0).getMusic());
            if(genreMusic.get(entry.getKey()).size() >= 2) {
                anslist.add(genreMusic.get(entry.getKey()).get(1).getMusic());
            }
        }
        // Integer list to primitive array
        int[] answer = anslist.stream().mapToInt(v->v).toArray();
        return answer;
    }
}
